lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -366,6 +364,135 @@ static LogicalResult verifyCircuitOp(CircuitOp circuit) {
+      if (!innerRef)
+        return nla.emitOpError()
+               << "the instance path can only contain inner sym reference"
+               << ", only the leaf can refer to a module symbol";
+
+      if (expectedModuleName && expectedModuleName != innerRef.getModule())
+        return nla.emitOpError()
+               << "instance path is incorrect. Expected module: "
+               << expectedModuleName
+               << " instead found: " << innerRef.getModule();
+      InstanceOp instOp =
+          dyn_cast_or_null<InstanceOp>(instanceSyms.getOpIfExists(innerRef));
+      if (!instOp)
+        return nla.emitOpError()
+               << " module: " << innerRef.getModule()
+               << " does not contain any instance with symbol: "
+               << innerRef.getName();
+      expectedModuleName = instOp.moduleNameAttr().getAttr();
+      AnnotationSet annos(instOp);
+      bool instFound = hasNonLocal(annos, nlaName);
+      if (!instFound) {
+        auto diag = nla.emitOpError()
+                    << " instance with symbol: " << innerRef
+                    << " does not contain a reference to the NonLocalAnchor";
+        diag.attachNote(instOp.getLoc()) << "the instance was defined here";
+        return failure();
+      }
+    }
+    // The instance path has been verified. Now verify the last element.
+    auto leafRef = namepath[namepath.size() - 1];
+    bool leafFound = false;
+    if (auto innerRef = leafRef.dyn_cast<hw::InnerRefAttr>()) {
+      auto *rec = leafInnerSyms.getRecordIfExists(innerRef);
+      if (!rec)
+        return nla.emitOpError()
+               << " operation with symbol: " << innerRef << " was not found ";
+      if (expectedModuleName != innerRef.getModule())
+        return nla.emitOpError()
+               << "instance path is incorrect. Expected module: "
+               << expectedModuleName
+               << " instead found: " << innerRef.getModule();
+
+      if (auto mod = dyn_cast_or_null<FModuleLike>(rec->op)) {
+        auto annos = AnnotationSet::forPort(mod, rec->portIdx);
+        leafFound = hasNonLocal(annos, nlaName);
+      } else if (rec->op) {
+        AnnotationSet annos(rec->op);
+        leafFound = hasNonLocal(annos, nlaName);
+      }
+      if (!leafFound) {
+        auto diag = nla.emitOpError()
+                    << " operation with symbol: " << innerRef
+                    << " does not contain a reference to the NonLocalAnchor";
+        diag.attachNote(rec->op->getLoc()) << "the symbol was defined here";
+        return failure();
+      }
+    } else if (expectedModuleName !=
+               leafRef.cast<FlatSymbolRefAttr>().getAttr()) {
+      // This is the case when the nla is applied to a module.
+      nla.emitOpError() << "instance path is incorrect. Expected module: "
+                        << expectedModuleName << " instead found: "
+                        << leafRef.cast<FlatSymbolRefAttr>().getAttr();
+      return failure();
```

Comment:
can these two functions also be updated to `return diag.attachNote()` and `return nla.emitOpError()`?

Reply:
Fixed, Thanks for spotting it.